### ISSUE TITLE
Slow full-backup on deployment with millions of small files 

### ISSUE THREAD

**Issue by MannerMan** [pgbackrest/pgbackrest#1149]
Please provide the following information when submitting an issue (feature requests or general comments can skip this):

1. pgBackRest version: **2.28**

2. PostgreSQL version: **9.5**

3. Operating system/version: **CentOS 7**

4. Did you install pgBackRest from source or from a package? **Package**

5. Please attach the following as applicable:
* `pgbackrest.conf`
```
[utv-db10]
pg1-path=/var/lib/pgsql/9.5/data

[global]
process-max=4
repo1-path=/
repo1-retention-full=2
repo1-cipher-type=none
repo1-s3-bucket=utv-db10
repo1-s3-endpoint=svc-filestor.lab.int
repo1-s3-host=utv-db10.svc-filestor.lab.int
repo1-s3-verify-ssl=n
repo1-s3-key=svc-filestor
repo1-s3-key-secret=<redacted>
repo1-s3-region=us-east-1
repo1-type=s3
compress-type=zst
start-fast=y
buffer-size=8388608

[global:archive-push]
compress-level=3
```

7. Describe the issue:

(This is a follow-up from issue #1118)

I'm evaluating pgbackrest as a replacement for my companys current backup system WAL-E. We have a schema-level style of sharding our tenants, where each of our database servers hosts 4000 customer schemas, spread over 16 databases. This has worked great for scalability, but presents a challenge for many backup systems - since this layout results in a lot of
files in postgresql data directory.

Output of `ls -1RL /var/lib/pgsql/9.5/data/ | wc -l`;
`6602805`

So above 6 million files in pgdata, total size is around 77gb. 

While we previously had problems with pgbackrest memory-usage, when that issue was quickly resolved we ran into a secondary problem; full-backups taking a very long time to complete. On the server mentioned above, performing a full backup took over 10 hours. The root cause was identified by pgbackrest developers as pgbackrest processing+uploading each file individually in pgdata dir, causing a large overhead when targeting S3 storage.

It was suggested in that issue (#1118) to employ a similar technique as WAL-E, i.e. compress several small files into a single compressed tarball before uploading, to avoid too much overhead. It was agreed that the implementation of such a feature should be tracked in a separate issue, which would be the one created here. With WAL-E, a full backup of the same server can be done in about 70 minutes.

I can help with testing any potential fixes for this problem on our lab environment. Please let me know if I can provide any more information or assist in other ways.

**Comment by MannerMan** [comment]
Just as a side-note; we tested doing a (partial with `--db-include`) restore using pgbackrest on our setup; it took 32h. It appears the restore process is single-threaded, further pushing the time to complete when dealing with so many files.

**Comment by sfrost** [comment]
If you have process-max set appropriately then pgbackrest will absolutely use as many parallel processes as you allow it to.  On the system where you're doing the restore, can you share the pgbackrest.conf file and the command that you're executing..?

**Comment by MannerMan** [comment]
@sfrost A bit weird that it took twice the time of a backup then, although it could be our storage being slow at writes as well.
If using `--db-include`, will pgbackrest still download all data-files and just zero out the ones not needed, or only download the files it actually requires? Since we in 99% of the cases just want to restore a single schema, we hoped the `--db-include` option could get us reasonable restore times.

The recovery command used:
`pgbackrest restore --stanza=utv-db10 --pg1-path=/pgdata/ --archive-mode=off --db-include=schemas_001 --type=time --target='2020-09-04 09:00:00 CEST' --log-level-console=info`
(our databases are confusingly named schemas_xxx)

pgbackrest.conf:
```
[utv-db10]
pg1-path=/var/lib/pgsql/9.5/data

[global]
process-max=4
repo1-path=/
repo1-retention-full=2
repo1-cipher-type=none
repo1-s3-bucket=utv-db10
repo1-s3-endpoint=svc-filestor.lab.int
repo1-s3-host=utv-db10.svc-filestor.lab.int
repo1-s3-verify-ssl=n
repo1-s3-key=svc-filestor
repo1-s3-key-secret=<redacted>
repo1-s3-region=us-east-1
repo1-type=s3
start-fast=y
compress-type=lz4
compress-level=1
buffer-size=1MB

[global:archive-push]
compress-level=3 
```

**Comment by MannerMan** [comment]
Trying another restore test;
I bumped the server to 8 cores, and passed in `--process-max=8` from command-line to be sure. I can observe 8 pgbackrest child-processes, so it does indeed run in parallel. CPU load distribution seems very concentrated to the single parent-process though.

![Screenshot from 2020-09-08 10-17-09](https://user-images.githubusercontent.com/1773068/92451330-dcdf7880-f1bc-11ea-9ae0-3277e99fa062.png)

I'll report back the results of this run when it finishes.

**Comment by dwsteele** [comment]
Seems like a lot of CPU for the main process -- I wonder if there is some new bottleneck that's causing problems here.

**Comment by MannerMan** [comment]
Restore is still running. I did a `perf` recording/profiling with
`perf record -F 99 -p 2866 --call-graph dwarf sleep 20`

Report here, if it helps:
[pgbackrest.txt](https://github.com/pgbackrest/pgbackrest/files/5192919/pgbackrest.txt)

**Comment by sfrost** [comment]
pgbackrest won't download files it doesn't need, no, it's just going to look at the manifest and put the zero'd files in place to support WAL replay (note that all WAL will still need to be replayed, which will include WAL that is touching other databases, but we don't have any control over that today really since that's just PG doing it and we don't do any WAL filtering, but that's all post-pgbackrest-restore anyway).

I'm guessing this is, again, a case where you've got bazillions of tiny files..?  So the main process is having trouble keeping the workers fed fast enough with work to do.  I imagine there's something we could do to work on improving that but I'm not sure that we're really going to be able to put any priority on it any time soon...  it's not a problem we run into very commonly and it's not a terribly good design due to exactly these kinds of issues...

**Comment by MannerMan** [comment]
@sfrost Yes, this is a restore of the instance with 6.6 million files. Unfortunately we're kind of stuck with this data-layout since it's a very large and active software system (150+ database servers). The suggested fix was to have pgbackrest package chunks of files into tar archives (or similar) and work with that to avoid having to process so many files individually.

Is it possible to discuss financial compensation to have pgbackrest support this use-case? Pgbackrest webpage seems to suggest Crunchy Data can provide support, should we turn to them?

**Comment by sfrost** [comment]
I'm not sure that it's quite so simple...  Using tar files is a nice idea to reduce the number of files that get stored, and it might help out with the "restore a single full backup", but the issue is going to be incremental restores since it's not clear how efficient it would be to try and pull out a single file from the middle of a tar file, if that's the only file that's needed for the restore from an incremental backup.

I've also reached out to you directly to chat further regarding prioritization.

**Comment by MannerMan** [comment]
@sfrost I'm guessing you sent an email to my public address here at GitHub? Unfortunately that address was decommissioned some time ago and I had forgot to update GitHub, sorry for that! Can I drop you an email to your address here from my work to get a conversation going?

**Comment by sfrost** [comment]
Sure, probably simplest to reach out to stephen@crunchydata.com

Thanks!

**Comment by Tostino** [comment]
You all may already be aware, but the [.dar](https://en.wikipedia.org/wiki/Dar_(disk_archiver)) format seems like it would be better suited to this use than .tar due to the seeking issues mentioned. Not sure what other issues it would bring with it though.

**Comment by dwsteele** [comment]
@MannerMan It was pretty clear from your perf report that the `memmove()` in `lstRemoveIdx()` was using up a lot of time so we optimized that code in PR #1191. Not only was the memmove() slow but it also appeared to do bad things to the CPU caches, perhaps because the source and destination are so heavily overlapped.

This optimization more than doubled the speed of my 1 million file test. The performance degradation in lstRemoveIdx() was pretty linear so this may well have a much larger impact on your backup/restore.

If you could give it a try we would appreciate it!

**Comment by MannerMan** [comment]
@dwsteele Wow, that sounds fantastic! I will do a test ASAP :+1:

**Comment by MannerMan** [comment]
Update here; I'm doing backup/restore tests to compare performance on 3.31dev vs 3.30, and I will test to see how well we can scale performance by increasing `process-max`. Since we have so many files it takes quite a lot of time, but I'll report back the status here as soon as I have some numbers :)

**Comment by dwsteele** [comment]
Cool! You built from the PR branch, right? This code is not in master yet.

**Comment by MannerMan** [comment]
@dwsteele Yes, I built it from the PR branch! :+1: We're having some problems with wal archiving timeouts, broken TCP connections and similiar which have terminated the backup some hours in. Currently troubleshooting this to see if it's something with our test setup or something with the new release. It's possible the new version, in being more efficient/parallel overloads our S3 backend (minio). Since the failures take quite a long time to occur, it's going a bit slow at the moment, but I'll report back my findings here as soon as I have some :)

**Comment by dwsteele** [comment]
I don't think this PR alone could have cause any networks issues, but it is always possible that something else previously committed has. It might be good to start with a low/default process-max.

**Comment by MannerMan** [comment]
@dwsteele 
We finally were able to figure out what was causing problems, we had this issue of pgbackrest complaining about wal-archiving timeouts during backup. Since we were running with `--log-level-console=warn` (we have so many files, `info` really floods the console), those timeouts appeared to happen randomly during the backup process. Actually it was just at the end of the backup, but we figured that out later :) We setup extensive monitoring/metric collection of the systems to try and track this issue down. That's when I noticed that we could not do a `pgbackrest check`, it just hung/got stuck. Some searching later we found https://github.com/pgbackrest/pgbackrest/issues/645#issuecomment-450230669 in which it turned out we had to set `MINIO_DOMAIN` to resolve this. And this option also fixed the archiving 'timeout' (it was not an actual timeout, pgbackrest simply could not find the wal files in the bucket without the `MINIO_DOMAIN` option set).

We can now do complete backups and it's looking very promising. I will have some number to share within the day or so!

**Comment by dwsteele** [comment]
@MannerMan Yeah, that makes sense. You can also use https://pgbackrest.org/configuration.html#section-repository/option-repo-s3-uri-style if you don't want to or can't use `MINIO_DOMAIN`. That option was introduced after the issue you were looking at.

**Comment by MannerMan** [comment]
> @MannerMan Yeah, that makes sense. You can also use https://pgbackrest.org/configuration.html#section-repository/option-repo-s3-uri-style if you don't want to or can't use `MINIO_DOMAIN`. That option was introduced after the issue you were looking at.

Oh, that's great - we'll be sure to set that option :+1: 

I finally have some performance numbers to share! These results are **not** comparable with the previous ones, since we in troubleshooting changed the setup quite considerable. We created a replica of the server we were testing and located it on very fast locally attached SSD, rather than the shared SAN used by the primary. We also increased `--process-max` to `8` and setup a minio s3 server on the local SSD storage as a stanza repo as well. We then took all backups from the replica and sent them to the new stanza repo.

To get comparable numbers, we ran full and differential backups with this exact setup, swapping only the pgbackrest binary.

**3.30:** 
* full: ~11.7 hours (706 minutes)
* diff: 40 minutes

**3.31dev:**
* full: ~3.6 hours (219 minutes)
* diff: 30 minutes

Our metric graphs show the difference very clearly as well, there's a massive difference of IOPS utilization for for 3.31;

**3.30**
![Screenshot from 2020-10-15 07-46-48](https://user-images.githubusercontent.com/1773068/96082187-0de95200-0ebb-11eb-85f6-934322b76d38.png)

**3.31dev**
![Screenshot from 2020-10-15 07-47-22](https://user-images.githubusercontent.com/1773068/96082194-104bac00-0ebb-11eb-8c30-ca3a8959f5b7.png)

CPU and memory usage appears to be fairly unchanged.

Fantastic improvement! :smiley:

**Comment by dwsteele** [comment]
Nice! I'm a little shocked that a `memmove()` would result in such a performance issue. I knew it was not optimal when I wrote it, but 8 hours to do `memmove()` on a 60MB array is pretty amazing. It does correlate with my test results, though.

I'm not sure there's much more we can do without grouping small files together, but this is a great improvement.

**Comment by MannerMan** [comment]
> Nice! I'm a little shocked that a `memmove()` would result in such a performance issue. I knew it was not optimal when I wrote it, but 8 hours to do `memmove()` on a 60MB array is pretty amazing. It does correlate with my test results, though.
> 
> I'm not sure there's much more we can do without grouping small files together, but this is a great improvement.

Indeed, it's probably hard to runtimes down even further without doing some sort of grouping, still - 3.6 hours for a first-backup is not too bad :) I have two more tests in mind to look for bottlenecks;
* Performing a recovery operation
* We also want to upgrade the replica to postgres 12 with checksums activated and see if it affects run-times significantly

**Comment by sfrost** [comment]
You mentioned previously that these numbers aren't comparable to the prior numbers...  Which makes me wonder: how does pgbackrest, for this use-case, now compare against wal-e, with everything else being the same?

Also- have you tried with additional pgbackrest processes?  There seems to be a lot of CPU still available and it'd be interesting to see if additional processes helps or not, or if fewer processes hurt (or help, which might also happen).

Agreed that testing recovery and other parameters would certainly be useful to do too.

**Comment by MannerMan** [comment]
I have done much testing this week and this will be a big info-dump, sorry!

> You mentioned previously that these numbers aren't comparable to the prior numbers... Which makes me wonder: how does pgbackrest, for this use-case, now compare against wal-e, with everything else being the same?

I can give it a spin, although right now I'm focusing on testing pgbackrest backup and restore

> Also- have you tried with additional pgbackrest processes? There seems to be a lot of CPU still available and it'd be interesting to see if additional processes helps or not, or if fewer processes hurt (or help, which might also happen).

I have now run backup with `process-max=4` and `process-max=16` (server has 8 cores)

With 4 processes, the runtime ended up at 4.6 hours for a full backup. So around one hour slower than with 8 processes.

With 16 processes, the runtime ended up at 4.3 hours for a full backup. So also slower than 8. pgbackrest also ended up using up a lot of memory at the end of the backup for the 16 process run
![Screenshot from 2020-10-21 15-13-12](https://user-images.githubusercontent.com/1773068/97012417-1f2fff80-1548-11eb-8367-7826fa7a6dac.png)
, but did complete successfully;
![Screenshot from 2020-10-23 15-01-27](https://user-images.githubusercontent.com/1773068/97006806-b002dd00-1540-11eb-995c-14d06f60cae0.png)
I did a perf while it was eating memory there at the end: [perf_during_backup_16c_high_memory.txt](https://github.com/pgbackrest/pgbackrest/files/5429608/perf_during_backup_16c_high_memory.txt)
I actually did a gdb/gcore dump while it was in this state as well, but I'm not sure I can share that since it might contain sensitive data..

One interesting thing when doing backups is the cpu utilization appears to be very 'jumpy', pgbackrest uses all 20-30% of cores for a while but then go back to using just one core at 100% and cycles like that.
**CPU Graph;**
![Screenshot from 2020-10-22 07-55-00](https://user-images.githubusercontent.com/1773068/97007247-6c5ca300-1541-11eb-9935-c7636f1f5667.png)


**Htop during 20-30% all core load;**
![Screenshot from 2020-10-19 09-31-48](https://user-images.githubusercontent.com/1773068/97007281-77173800-1541-11eb-80e9-b5751eaceca2.png)

**Htop during 100% load on a single core;**
![Screenshot from 2020-10-19 09-34-01](https://user-images.githubusercontent.com/1773068/97007292-7aaabf00-1541-11eb-9625-1e40756401cd.png)


I performed perf recordings both when pgbackrest was using all cores and when it was using only one, if it can be of any use. This was from a regular `process-max=8` run.
[perf_during_backup_load_all_core.txt](https://github.com/pgbackrest/pgbackrest/files/5429851/perf_during_backup_load_all_core.txt)
[perf_during_backup_load_one_core.txt](https://github.com/pgbackrest/pgbackrest/files/5429853/perf_during_backup_load_one_core.txt)
 
> Agreed that testing recovery and other parameters would certainly be useful to do too.

I have now also tried a recovery operation, and there still seems to be some potential for improvements here. Both 3.30 and 3.31dev took over 20 hours and used up ~5.3gb of memory. Pgbackrest puts 100% load on a single core while recovery is running, with little activity on the remaining seven for the remainder of the operation.
**Htop during recovery;**
![Screenshot from 2020-10-21 13-06-25](https://user-images.githubusercontent.com/1773068/97007614-f73d9d80-1541-11eb-93a0-13750457d588.png)

**CPU/Memory/Disk/Network graphs;**
![Screenshot from 2020-10-22 07-32-24](https://user-images.githubusercontent.com/1773068/97007760-31a73a80-1542-11eb-9269-0fb23985be46.png)
![Screenshot from 2020-10-22 07-32-48](https://user-images.githubusercontent.com/1773068/97007765-3370fe00-1542-11eb-8aab-951dff4090ea.png)


Recovery seems split into two segments, the first one from 22:00 to around 07:00 in the graphs, where there are network and some disk activity. Then a second segment from 07:00 and on-wards where there is no network activity and just a tiny amount of disk.

In the second segment, pgbackrest console logs a huge amount of 0B files being restored;
![Screenshot from 2020-10-21 15-13-12](https://user-images.githubusercontent.com/1773068/97012464-28b96780-1548-11eb-8242-611f48a0dc79.png)

(I guess these are empty tables, indexes etc..)

I took out perf reports both during the first and second segments of recovery;
[perf_during_recovery_segment1.txt](https://github.com/pgbackrest/pgbackrest/files/5429678/perf_during_recovery_segment1.txt)
[perf_during_end_of_recovery_segment2.txt](https://github.com/pgbackrest/pgbackrest/files/5429694/perf_during_end_of_recovery_segment2.txt)

Let me know if I can provide any further information that might be of use!

**Comment by MikeN123** [comment]
I would just like to mention that backing up many small files to cloud storage also has cost implications. I thought this was already mentioned somewhere, but can't find it and this ticket seems a good place to mention it.

As an example, I have 30k files and a 1 GB full backup after compression. Costs for this in Azure are about € 0.0166/month for the storage, but also € 0.1368 just to create the backup, i.e., the creation of the backup is 8 times more expensive than the storage. Depending on your backup size vs. the number of files, this means that the operations may cost much more than the actual storage. With our database sizes this is probably not an issue, but I do think it may drive people to other solutions in certain cases.

I fully understand it is probably not easy to fix, just mentioning this for reference.

**Comment by dwsteele** [comment]
We are definitely aware of the cost implications of many small files in a backup. It's yet another argument for better small file support.

We have definitely discussed this aspect before, but good to have it noted here as well.

**Comment by FrederickCoene** [comment]
Hello, 

I'm also experiencing the same issue. 
Our database is around 2TB in size. We store pgbackrest data on a CIFS volume.
It takes more than 2.5 days to create a full backup, due to the large amount of small files.

What is the status of this issue?

**Comment by dwsteele** [comment]
We are actively working on this, but it probably won't be ready for the March release. How many files do you have?

Are restores just as much of a problem?

**Comment by FrederickCoene** [comment]
pgBackRest 2.33

pgbackrest.conf
`
[global]
repo1-path=/var/lib/pgbackrest
repo1-retention-full=1
repo1-retention-diff=2
repo1-type=cifs
repo1-cipher-pass=<redacted>
repo1-cipher-type=aes-256-cbc
start-fast=y
process-max=12

[global:archive-push]
compress-level=3
process-max=4

[main]
pg1-path=/var/lib/pgsql/13/data/
`

disks
`df -h`
`
/dev/md1                              98G   42G   52G  45% /var/log
/dev/md126                           894G   70G  825G   8% /wal
/dev/md127                           3.5T  1.9T  1.7T  54% /var/lib/pgsql
//<redacted>/backup  2.0T  883G  1.1T  45% /var/lib/pgbackrest (cifs volume)
`

`ls -1RL /var/lib/pgsql/13/data/ | wc -l`
`4400344`

We have around 4.4Mio files.

the full backup is done with 12 processes but takes +/-2.5 days.
Logs of the backup (it's currently still busy):
`
-------------------PROCESS START-------------------
2022-02-06 06:30:01.554 P00   INFO: backup command begin 2.33: --exec-id=1123836-57d67cbd --pg1-path=/var/lib/pgsql/13/data --process-max=12 --repo1-cipher-pass=<redacted> --repo1-cipher-type=aes-256-cbc --repo1-path=/var/lib/pgbackrest --repo1-retention-diff=2 --repo1-retention-full=1 --repo1-type=cifs --stanza=main --start-fast --type=full
2022-02-06 06:30:03.536 P00   INFO: execute non-exclusive pg_start_backup(): backup begins after the requested immediate checkpoint completes
2022-02-06 06:30:04.238 P00   INFO: backup start archive = 00000001000004370000005E, lsn = 437/5E000028
2022-02-06 10:52:00.218 P02   INFO: backup file /var/lib/pgsql/13/data/base/16385/39266084 (1GB, 0%) checksum 0c7cdac797f70e159f89b1bf172e8ab535b486c2
2022-02-06 10:52:00.371 P01   INFO: backup file /var/lib/pgsql/13/data/base/16385/39266084.1 (1GB, 0%) checksum 4722aa2558c8ed63dd91e288e8ac64446b4f5092
2022-02-06 10:52:05.333 P12   INFO: backup file /var/lib/pgsql/13/data/base/16385/39263905 (1GB, 0%) checksum 0a8e32734b92312ecaed06eb97abc902fc7955d6
2022-02-06 10:52:05.342 P07   INFO: backup file /var/lib/pgsql/13/data/base/16385/39263911 (1GB, 0%) checksum 97ee101f59b728a65553e10963adbaa246fcf1f8
2022-02-06 10:52:05.365 P10   INFO: backup file /var/lib/pgsql/13/data/base/16385/39263905.2 (1GB, 0%) checksum 04027b24462ce1097929f56afaa2e41ed580ed0a
2022-02-06 10:52:05.532 P03   INFO: backup file /var/lib/pgsql/13/data/base/16385/39263911.4 (1GB, 0%) checksum b2fa33f4c6abdba232dc5533150b516648b1514a
2022-02-06 10:52:05.866 P08   INFO: backup file /var/lib/pgsql/13/data/base/16385/39263905.4 (1GB, 0%) checksum 8a685dc150acedb712db6ba86b2f2e1af6c40419
2022-02-06 10:52:06.234 P09   INFO: backup file /var/lib/pgsql/13/data/base/16385/39263905.3 (1GB, 0%) checksum 63370cf515c6a3abe8537bf05dbb1b813ca63ea0
2022-02-06 10:52:06.654 P06   INFO: backup file /var/lib/pgsql/13/data/base/16385/39263911.1 (1GB, 0%) checksum 7ffdd12262b41c25f773f81fd001282e874f310b
2022-02-06 10:52:07.145 P11   INFO: backup file /var/lib/pgsql/13/data/base/16385/39263905.1 (1GB, 0%) checksum 8964b33d9f01729990c714654e7ad5b41bfa6af9
2022-02-06 10:52:07.406 P05   INFO: backup file /var/lib/pgsql/13/data/base/16385/39263911.2 (1GB, 1%) checksum bd5e941f5724f2ee1e8adc55ac87e1812f987470

...

2022-02-06 14:40:56.655 P03   INFO: backup file /var/lib/pgsql/13/data/base/16385/26010223 (368KB, 94%) checksum 3f8cd5c764669b3dcfac1382193581f088c99d99
2022-02-06 14:40:56.658 P10   INFO: backup file /var/lib/pgsql/13/data/base/16385/26010221 (368KB, 94%) checksum f95c27a3bd8363a2b0b25b03712982c4f6e4126b

...

2022-02-06 14:40:56.655 P03   INFO: backup file /var/lib/pgsql/13/data/base/16385/26010223 (368KB, 94%) checksum 3f8cd5c764669b3dcfac1382193581f088c99d99
2022-02-06 14:40:56.658 P10   INFO: backup file /var/lib/pgsql/13/data/base/16385/26010221 (368KB, 94%) checksum f95c27a3bd8363a2b0b25b03712982c4f6e4126b

...

2022-02-07 08:49:13.522 P05   INFO: backup file /var/lib/pgsql/13/data/base/16385/25546426 (8KB, 99%) checksum 2f1e1ea07bb857cc377f0110419ae6d35e7a8f91
2022-02-07 08:49:13.524 P01   INFO: backup file /var/lib/pgsql/13/data/base/16385/25546427 (8KB, 99%) checksum 165d3b281a5a62d2b645e4cc26eefd41ee378f73

...

2022-02-08 16:00:50.959 P05   INFO: backup file /var/lib/pgsql/13/data/base/16385/17327692_init (0B, 99%)
2022-02-08 16:00:50.961 P03   INFO: backup file /var/lib/pgsql/13/data/base/16385/17327696_init (0B, 99%)

...

2022-02-08 16:01:19.460 P06   INFO: backup file /var/lib/pgsql/13/data/base/16385/17277049_init (0B, 99%)
2022-02-08 16:01:19.462 P02   INFO: backup file /var/lib/pgsql/13/data/base/16385/17276602_init (0B, 99%)
`


The restore is actually really fast. I use 50 processes to restore and it's done in less than 3 hours.

**Comment by dwsteele** [comment]
Thank you for the info! What we are working on should help here quite a bit.

Having said that, this still seems like a *really* long time for four million files. I suspect storage speed is a major factor here. When writing out the zero-length files how long is the interval between files per process? So, what did process 6 complete after:

**Comment by FrederickCoene** [comment]
Would it help if we would move to Amazon S3 storage (instead of using CIFS storage)?

**Comment by dwsteele** [comment]
It's taking about a second per file which definitely explains why it is taking so long.

> Would it help if we would move to Amazon S3 storage (instead of using CIFS storage)?

Probably not. Per-file creation cost in S3 is fairly high. Object stores are the main reason we are working on small-file support. CIFS/Posix are generally much faster, but it all depends on the storage.

**Comment by FrederickCoene** [comment]
Ok, thanks for working on this. 
This will be a big improvement for us.

Our CIFS volume is hosted by Hetzner.

**Comment by MannerMan** [comment]
I want to chime in with a thanks for working on this, we have over 200 postgres clusters with 4M+ files in every cluster, and backing it up is a bit of a hassle at the moment. Impressively though pgbackrest does work since 3.31+, even if it takes some time.

Will it be possible to activate/use the upcoming solution/feature 'on the fly' or will it require a new stanza initialization?

**Comment by dwsteele** [comment]
> Will it be possible to activate/use the upcoming solution/feature 'on the fly' or will it require a new stanza initialization?

No need to create a new repo (or even a new full backup), though of course backups taken with this method won't work with older versions of pgbackrest.

**Comment by dwsteele** [comment]
I have posted a WIP patch for small file bundling feature: https://github.com/pgbackrest/pgbackrest/pull/1655

**Comment by MannerMan** [comment]
> I have posted a WIP patch for small file bundling feature: #1655

Awesome! Is it in a good enough state for testing right now? I can assist in that case and try it out on one of our dev clusters, if that might be of help?

**Comment by dwsteele** [comment]
> Is it in a good enough state for testing right now?

This is fully functional as far as we know. Certainly good enough for testing.

**Comment by dwsteele** [comment]
And yes, it would be a big help if you tested so we can get some real-world numbers.

**Comment by MannerMan** [comment]
So I have some initial number to share. 

First I'd like to give some background on our setup;
We self-host everything, and our S3 storage is provided by Minio. We have heavily optimized for the current implementation of backing up millions of tiny files. Clustered Minio instances split incoming files into several smaller chunks, which caused enormous overhead due to each tiny file taking up 4kb in the end (block size). A 11GB backup set size ended up taking a ridiculous ~350gb on a 32 node Minio cluster.

To work around this, we have setup several independent single/standalone instances of Minio, configured with a xfs block size of 1kb. Each database server sends its backups to one of these instances. This way we avoid a lot of network, chunking and block size overhead for each request compared to a clustered setup. The load on the individual Minio instances becomes quite severe during backup runs, but it has been working well. The same 11GB backup set claims around 14 gb on disk with this layout.

Since we have optimized this way, the differences in _backup size and time_ was minor with and without the bundle feature. 
On a 127GB postgres cluster with 4M files, a regular backup performed on 2.36 took ~2.6 hours. With 2.38+bundle, (100MiB bundle-size) it took ~2.7 hours. However, the difference in load at the Minio side is quite amazing, we went from 600 req/sec to around 10 req/sec with 2.38+bundle during backup. And of course the number of files produced:

2.36: 4033673
2.38: 785

When running multiple backups at once during weekends, we get longer backup times on 2.36 since Minio gets overloaded, generally backups take around ~4 hours then. I would expect 2.38 to improve this greatly, since the load on Minio is barely noticeable when using bundle feature :) It's also very likely that when using a clustered Minio instance as backend, 2.38 would have performed much better than 2.36, since the overhead for each request would have been much larger. We will migrate to a clustered setup if bundle becomes generally available as a feature, for ease management and high-availability.

I'll see if I can get a restore test going next week! I'll also play around some more with the bundle-size option and max-process.

Please let me know if any other test or configuration might be of use!

**Comment by dwsteele** [comment]
Excellent, thanks for doing that! These are really interesting numbers -- it's amazing how much you have optimized this case.

> if bundle becomes generally available as a feature

This will be a feature, but I doubt it will be GA for the 2.38 release. We'll keep it as an internal feature for at least another release.

**Comment by MannerMan** [comment]
> Excellent, thanks for doing that! These are really interesting numbers -- it's amazing how much you have optimized this case.

No problem! Heh, yes it has been an interesting challenge, we currently have 2 haproxy servers doing 16000 S3 requests/sec at peak when we back up our production dbs. We have 32 standalone Minio servers, each database cluster gets it minio s3 host set with the formula `(${internal_cluster_number} % 32)+1`. I should probably write a blogpost or something :sweat_smile:

> This will be a feature, but I doubt it will be GA for the 2.38 release. We'll keep it as an internal feature for at least another release.

Fantastic! We'll benefit greatly by this feature!

**Comment by dwsteele** [comment]
We have just added a `bundle-limit` option to restrict which files are added to bundles by size. See PR #1662.

This should allow for more fine-tuning and it makes resume work again.

**Comment by MannerMan** [comment]
> We have just added a `bundle-limit` option to restrict which files are added to bundles by size. See PR #1662.
> 
> This should allow for more fine-tuning and it makes resume work again.

Cool! I'll see if I can play around with that setting as well and post the results here.

I did some restore testing with backups from 2.36 and 2.38dev+bundle for our setup. Keep in mind that this using the same setup as described above, optimized for the many-small-files scenario. Restoring the 127gb postgres cluster that we previously tested as backup target, we got the following results;
2.36: 1h8m
2.38+bundle: 50m

So 2.38 was noticeably faster for us. If we had been using a remote or clustered S3 setup, the differences would have been much larger.

**Comment by MannerMan** [comment]
Some more testing with options;

```
bundle-size=1024MiB
bundle-limit=128MiB
```
Total backup time: 1h14m
Files produced: 286

```
bundle-size=20MiB
bundle-limit=2MiB
```
Total backup time: 2h12mQ
Files produced: 3832

So for us, using a larger bundle-size than default is beneficial.

One thing I noticed was that the bundle files does not become exactly the size configured, or even very close to it. For example here is an excerpt from a run with bundle-size: 1024mb. Not an issue though, I guess this might be because of compression?
![Screenshot from 2022-02-23 09-06-47](https://user-images.githubusercontent.com/1773068/155281739-afe112ee-a5e3-490d-949a-b4180cc45c57.png)

Is there any other scenario you would like me to test with our setup?

**Comment by dwsteele** [comment]
> So for us, using a larger bundle-size than default is beneficial.

It was definitely the intention that special cases like yours would need some tuning, but the defaults don't look too bad, either.

> One thing I noticed was that the bundle files does not become exactly the size configured, or even very close to it.

Yeah, the size is before compression. The original goal was to try to get the final file size to be `bundle-size` but now I don't think that is a worthwhile goal because it involves a lot of complexity and might end up in retries being stuck with a lot of work when we implement page-level incremental backups. We have greatly reduced the number of files being written which is the important part.

> Is there any other scenario you would like me to test with our setup?

Nothing I can think of right now. When we have the restore optimizations done I'll let you know.

**Comment by dwsteele** [comment]
Restore optimizations have been added in PR #1683. This will greatly reduce the number of GETs that are done against the object store (though the same total about of data will be read).

**Comment by MannerMan** [comment]
> Restore optimizations have been added in PR #1683. This will greatly reduce the number of GETs that are done against the object store (though the same total about of data will be read).

Great! I compiled the integration branch and re-ran some restore tests today;
```
3.36 no bundle:   ~ 1 hour 34 minutes
3.38dev:          ~ 50 minutes
3.39dev:          ~ 36 minutes
```

Looking very good! :smiley:

**Comment by dwsteele** [comment]
Yeah, that looks pretty good. Thanks!

I think we can close this issue now.

**Comment by MannerMan** [comment]
> Yeah, that looks pretty good. Thanks!
> 
> I think we can close this issue now.

Yup, any estimate when this feature will make it into a release?

**Comment by dwsteele** [comment]
I expect this to be in the May (2.39) release.

**Comment by MannerMan** [comment]
> I expect this to be in the May (2.39) release.

Thank you for the estimation, we'll make good use of this feature when it goes GA! :+1:
